library/src/hcc_detail/hipblas.cpp

```diff
@@ -342,10 +434,147 @@ int m, int n, int k,  const double *alpha, const double *A, int lda, long long b
+        B, ldb, bsb_int, 
+        beta, 
+        C, ldc, bsc_int, 
+        batchCount));
 }
 
 #ifdef __cplusplus
 }
 #endif
+
+template<typename T>
+rocblas_status 
+gemm_template(rocblas_handle handle,
+    rocblas_operation transA, rocblas_operation transB,
+    rocblas_int m, rocblas_int n, rocblas_int k,
+    const T *alpha,
+    const T *A, rocblas_int lda,
+    const T *B, rocblas_int ldb,
+    const T *beta,
+    T *C, rocblas_int ldc);
+
+template<>
+rocblas_status
+gemm_template<float>(rocblas_handle handle,
+    rocblas_operation transA,
+    rocblas_operation transB,
+    rocblas_int M, rocblas_int N, rocblas_int K,
+    const float *alpha,
+    const float *A, rocblas_int lda,
+    const float *B, rocblas_int ldb,
+    const float *beta,
+    float *C, rocblas_int ldc)
+{
+    return rocblas_sgemm(handle, transA, transB, M, N, K, alpha, A, lda, B, ldb, beta, C, ldc);
+}
+
+template<>
+rocblas_status
+gemm_template<double>(rocblas_handle handle,
+    rocblas_operation transA,
+    rocblas_operation transB,
+    rocblas_int M, rocblas_int N, rocblas_int K,
+    const double *alpha,
+    const double *A, rocblas_int lda,
+    const double *B, rocblas_int ldb,
+    const double *beta,
+    double *C, rocblas_int ldc)
+{
+    return rocblas_dgemm(handle, transA, transB, M, N, K, alpha, A, lda, B, ldb, beta, C, ldc);
+}
+
+template<typename T>
+hipblasStatus_t 
+hipblasGemmBatched_template(hipblasHandle_t handle,  
+        hipblasOperation_t transa, hipblasOperation_t transb,
+        int m, int n, int k,
+        const T *alpha,
+        const T * const A[], int lda, 
+        const T * const B[], int ldb, 
+        const T *beta, 
+        T * const C[], int ldc, int batchCount)
+{
+    if (batchCount < 0 || m < 0 || n < 0 || k < 0 || lda < 0 || ldb < 0 || ldc < 0) 
+    {
+        return HIPBLAS_STATUS_INVALID_VALUE;
+    }
+
+    T *hA[batchCount], *hB[batchCount], *hC[batchCount];
+    T h_alpha, h_beta;
+    // copy arrays from device to host
+    hipError_t err_A = hipMemcpy(hA, A, batchCount * sizeof(*A), hipMemcpyDeviceToHost);
+    hipError_t err_B = hipMemcpy(hB, B, batchCount * sizeof(*B), hipMemcpyDeviceToHost);
+    hipError_t err_C = hipMemcpy(hC, C, batchCount * sizeof(*C), hipMemcpyDeviceToHost);
+    if ((err_A != hipSuccess) || (err_B != hipSuccess) || (err_C != hipSuccess))
+    {
+        return HIPBLAS_STATUS_ALLOC_FAILED;
+    }
+
+    // copy alpha and beta to host if they are on device
+    if (rocblas_pointer_to_mode((void*)alpha) == rocblas_pointer_mode_device)
+    {
+        hipError_t err_alpha = hipMemcpy(&h_alpha, alpha, sizeof(T), hipMemcpyDeviceToHost);
+        hipError_t err_beta= hipMemcpy(&h_beta, beta, sizeof(T), hipMemcpyDeviceToHost);
+        if ((err_alpha != hipSuccess) || (err_beta != hipSuccess))
+        {
+            return HIPBLAS_STATUS_ALLOC_FAILED;
+        }
+    }
+    else
+    {
+        h_alpha = *alpha;
+        h_beta= *beta;
+    }
+    for (int i = 0; i < batchCount; i++)
```

Comment:
Making a note: in the future we should check to see if the input matrices are strided, and call the strided API instead.